File: java/src/FrequencyTable.java
/* 
 * Reference arithmetic coding
 * Copyright (c) Project Nayuki
 * 
 * https://www.nayuki.io/page/reference-arithmetic-coding
 * https://github.com/nayuki/Reference-arithmetic-coding
 */


/**
 * A table of symbol frequencies. The table holds data for symbols numbered from 0
 * to getSymbolLimit()&minus;1. Each symbol has a frequency, which is a non-negative integer.
 * <p>Frequency table objects are primarily used for getting cumulative symbol
 * frequencies. These objects can be mutable depending on the implementation.
 * The total of all symbol frequencies must not exceed Integer.MAX_VALUE.</p>
 */
public interface FrequencyTable {
	
	/**
	 * Returns the number of symbols in this frequency table, which is a positive number.
	 * @return the number of symbols in this frequency table
	 */
	public int getSymbolLimit();
	
	
	/**
	 * Returns the frequency of the specified symbol. The returned value is at least 0.
	 * @param symbol the symbol to query
	 * @return the frequency of the symbol
	 * @throws IllegalArgumentException if the symbol is out of range
	 */
	public int get(int symbol);
	
	
	/**
	 * Sets the frequency of the specified symbol to the specified value.
	 * The frequency value must be at least 0.
	 * @param symbol the symbol to set
	 * @param freq the frequency value to set
	 * @throws IllegalArgumentException if the frequency is negative or the symbol is out of range
	 * @throws ArithmeticException if an arithmetic overflow occurs
	 */
	public void set(int symbol, int freq);
	
	
	/**
	 * Increments the frequency of the specified symbol.
	 * @param symbol the symbol whose frequency to increment
	 * @throws IllegalArgumentException if the symbol is out of range
	 * @throws ArithmeticException if an arithmetic overflow occurs
	 */
	public void increment(int symbol);
	
	
	/**
	 * Returns the total of all symbol frequencies. The returned value is at
	 * least 0 and is always equal to {@code getHigh(getSymbolLimit() - 1)}.
	 * @return the total of all symbol frequencies
	 */
	public int getTotal();
	
	
	/**
	 * Returns the sum of the frequencies of all the symbols strictly
	 * below the specified symbol value. The returned value is at least 0.
	 * @param symbol the symbol to query
	 * @return the sum of the frequencies of all the symbols below {@code symbol}
	 * @throws IllegalArgumentException if the symbol is out of range
	 */
	public int getLow(int symbol);
	
	
	/**
	 * Returns the sum of the frequencies of the specified symbol
	 * and all the symbols below. The returned value is at least 0.
	 * @param symbol the symbol to query
	 * @return the sum of the frequencies of {@code symbol} and all symbols below
	 * @throws IllegalArgumentException if the symbol is out of range
	 */
	public int getHigh(int symbol);
	
}

File: cpp/FrequencyTable.cpp
/* 
 * Reference arithmetic coding
 * Copyright (c) Project Nayuki
 * 
 * https://www.nayuki.io/page/reference-arithmetic-coding
 * https://github.com/nayuki/Reference-arithmetic-coding
 */

#include <stdexcept>
#include "FrequencyTable.hpp"

using std::uint32_t;


FrequencyTable::~FrequencyTable() {}


FlatFrequencyTable::FlatFrequencyTable(uint32_t numSyms) :
		numSymbols(numSyms) {
	if (numSyms < 1)
		throw std::domain_error("Number of symbols must be positive");
}


uint32_t FlatFrequencyTable::getSymbolLimit() const {
	return numSymbols;
}


uint32_t FlatFrequencyTable::get(uint32_t symbol) const  {
	checkSymbol(symbol);
	return 1;
}


uint32_t FlatFrequencyTable::getTotal() const  {
	return numSymbols;
}


uint32_t FlatFrequencyTable::getLow(uint32_t symbol) const  {
	checkSymbol(symbol);
	return symbol;
}


uint32_t FlatFrequencyTable::getHigh(uint32_t symbol) const  {
	checkSymbol(symbol);
	return symbol + 1;
}


void FlatFrequencyTable::set(uint32_t, uint32_t)  {
	throw std::logic_error("Unsupported operation");
}


void FlatFrequencyTable::increment(uint32_t) {
	throw std::logic_error("Unsupported operation");
}


void FlatFrequencyTable::checkSymbol(uint32_t symbol) const {
	if (symbol >= numSymbols)
		throw std::domain_error("Symbol out of range");
}


SimpleFrequencyTable::SimpleFrequencyTable(const std::vector<uint32_t> &freqs) {
	if (freqs.size() > UINT32_MAX - 1)
		throw std::length_error("Too many symbols");
	uint32_t size = static_cast<uint32_t>(freqs.size());
	if (size < 1)
		throw std::invalid_argument("At least 1 symbol needed");
	
	frequencies = freqs;
	cumulative.reserve(size + 1);
	initCumulative(false);
	total = getHigh(size - 1);
}


SimpleFrequencyTable::SimpleFrequencyTable(const FrequencyTable &freqs) {
	uint32_t size = freqs.getSymbolLimit();
	if (size < 1)
		throw std::invalid_argument("At least 1 symbol needed");
	if (size > UINT32_MAX - 1)
		throw std::length_error("Too many symbols");
	
	frequencies.reserve(size + 1);
	for (uint32_t i = 0; i < size; i++)
		frequencies.push_back(freqs.get(i));
	
	cumulative.reserve(size + 1);
	initCumulative(false);
	total = getHigh(size - 1);
}


uint32_t SimpleFrequencyTable::getSymbolLimit() const {
	return static_cast<uint32_t>(frequencies.size());
}


uint32_t SimpleFrequencyTable::get(uint32_t symbol) const {
	return frequencies.at(symbol);
}


void SimpleFrequencyTable::set(uint32_t symbol, uint32_t freq) {
	if (total < frequencies.at(symbol))
		throw std::logic_error("Assertion error");
	uint32_t temp = total - frequencies.at(symbol);
	total = checkedAdd(temp, freq);
	frequencies.at(symbol) = freq;
	cumulative.clear();
}


void SimpleFrequencyTable::increment(uint32_t symbol) {
	if (frequencies.at(symbol) == UINT32_MAX)
		throw std::overflow_error("Arithmetic overflow");
	total = checkedAdd(total, 1);
	frequencies.at(symbol)++;
	cumulative.clear();
}


uint32_t SimpleFrequencyTable::getTotal() const {
	return total;
}


uint32_t SimpleFrequencyTable::getLow(uint32_t symbol) const {
	initCumulative();
	return cumulative.at(symbol);
}


uint32_t SimpleFrequencyTable::getHigh(uint32_t symbol) const {
	initCumulative();
	return cumulative.at(symbol + 1);
}


void SimpleFrequencyTable::initCumulative(bool checkTotal) const {
	if (!cumulative.empty())
		return;
	uint32_t sum = 0;
	cumulative.push_back(sum);
	for (uint32_t freq : frequencies) {
		// This arithmetic should not throw an exception, because invariants are being maintained
		// elsewhere in the data structure. This implementation is just a defensive measure.
		sum = checkedAdd(freq, sum);
		cumulative.push_back(sum);
	}
	if (checkTotal && sum != total)
		throw std::logic_error("Assertion error");
}


uint32_t SimpleFrequencyTable::checkedAdd(uint32_t x, uint32_t y) {
	if (x > UINT32_MAX - y)
		throw std::overflow_error("Arithmetic overflow");
	return x + y;
}
